[COIN 275] Get account sequence and number from CosmosLikeAccount

gagbo opened the pull request:
The `cosmos::Account` data structure wasn't stored in the implementation class `core::CosmosLikeAccount`. The data structure has been added and is updated on each synchronization using the existing explorer method.

Tests have been added on `MediumXpubSynchronization`

Commit:
[CI] double timeout on tests

Commit:
Add CosmosAccount->getSequence() and getAccountNumber

Those methods are useful to build a transaction correctly

haammar-ledger commented on core/src/wallet/cosmos/CosmosLikeAccount.hpp:
```diff
@@ -150,6 +154,7 @@ namespace ledger {
 
+                                std::shared_ptr<cosmos::Account> _accData;
                                 std::shared_ptr<CosmosLikeKeychain> _keychain;
                                 std::string _accountAddress;
```
This can probably be removed and `_accData.address` used instead?

gagbo replied:
I'll check. I'm worried that the address is necessary for initial synchronization, and accData is not valid until after first sync :thinking:

haammar-ledger commented on core/src/wallet/cosmos/CosmosLikeAccount.hpp:
```diff
@@ -150,6 +154,7 @@ namespace ledger {
                         private:
                                 std::shared_ptr<CosmosLikeAccount> getSelf();
 
+                                std::shared_ptr<cosmos::Account> _accData;
```
I usually find it more readable to use full words for variables and functions, as long as they reasonably fit on the screen, but REALLY a personal taste matter so I won't insist on this one ^^

haammar-ledger commented on core/src/wallet/cosmos/CosmosLikeAccount.cpp:
```diff
@@ -600,20 +610,38 @@ namespace ledger {
                                 tx->setGas(request.gas);
                                 tx->setMemo(request.memo);
                                 tx->setMessages(request.messages);
-                                tx->setSequence(request.sequence);
```
`sequence` can maybe be removed from `CosmosLikeTransactionBuildRequest`

gagbo replied:
I'll make it optional instead then. If the live wants to set the sequence themselves we'll let them (in case they want to build multiple transactions "offline" before broadcasting, or keeping an internal sequence count or something)

gagbo replied:
So probably `request.sequence.empty() ? self->getSequence() : request.sequence`

Commit:
[review] remove redundant attribute

Making sure that the attributes are synchronized is better avoided.

Commit:
[review] allow TransactionBuildRequest to override sequence

haammar-ledger commented on core/src/wallet/cosmos/CosmosLikeAccount.cpp:
```diff
@@ -626,20 +627,22 @@ namespace ledger {
                 {
-                    if (!_accData) {
+                    // Assuming _accountData is only valid if accountNumber has been fetched
+                    if (!_accountData || _accountData->accountNumber.empty()) {
```
You mean `_accountData->sequence` here I think

haammar-ledger replied:
Oh nevermind I think I got the logic wrong: actually you test against `accountNumber` to check that the account is synchronized.

gagbo replied:
Not really, the `sequence` can be empty (maybe if the account never had any transaction), but I assumed that the accountNumber never is. It may be a wrong assumption now that I think about it, but I need to check that

gagbo replied:
Turns out I was wrong.

Commit:
Remove dead code

Commit:
Generate new random wallet names in tests

This should help avoid db conflicts resulting in false positive in CI

Commit:
Fix CosmosLikeAccount address being emptied on new accounts

If the account has no transaction the explorer may reset the address
field of the CosmosLikeAccount. This fixes this behaviour

Commit:
Remove useless validity check

Client code will be responsible of handling empty/nil values, and the
library should not try to be too smart about it ( or assume the explorer
implementation )
